### DISCUSSION TITLE
Print Typeface not rendering

### DISCUSSION THREAD

**Discussion by Kush-4826** [PDConSec/vsc-print#372] [Q&A] [answered]
The Type face that is selected in the VS Code is not applied to the prints. The prints appear in the browser default Serif font.

This is the Default Font Family Setting in the IDE
![image](https://github.com/PDConSec/vsc-print/assets/165514998/0581dc62-905f-461c-96d3-305a33ece529)

This is What the code looks like in the IDE
![image](https://github.com/PDConSec/vsc-print/assets/165514998/226c69f5-0b86-441d-9e2a-4856f5990420)

And this is what the Print looks like
![image](https://github.com/PDConSec/vsc-print/assets/165514998/3f5c9846-c846-48c0-bc64-2022d5a12c4f)

**Comment by ChaganKhagan** [comment]
Similar problem here, only I set my custom typeface in the VS Code but what's printed is the default Menlo.
Suppose this problem is either related to Non-System typefaces or when the fonts have format of .otf

**Comment by PeterWone** [comment]
Sorry to keep you both waiting, we have some build problems blocking release so the backlog has been building. First I'll get the build going, then the low hanging fruit, then I'll put some time into this one.

**Comment by PeterWone** [comment]
While I'm getting to this, would you please

1. Tell me what platform you're using (Win/Linux/MacOS).
2. Tell me what browser you're using (a Chromium derivative, Firefox, Safari or something else)
3. Disable `Print and Close` in your settings so the browser print dialog doesn't open when the print job is sent to the browser then trigger the problem and inspect the page to tell me what font-family is being specified for the incorrectly formatted text

If you can determine which CSS class is actually contributing the value in use, that would be super handy.

**Comment by ChaganKhagan** [comment]
Hi Peter!

Thanks for reaching to me.

I have been doing some very tiny experiments myself and this is what I have discovered. Hopefully it will be of any help.

First I’ll answer your questions.
1. I am using macOS 14 (Sonoma)
2. I am using Safari
3. This is the font settings I applied on my VS Code:
￼
4. This is the snapshot of the source code I took from Safari:
￼

I have confirmed that all four lines (font-family, font-size, line-height and padding bottom) are manually adjustable in situ and the changes will take effect realtime.

If I change the default font-family to another ’System' Font, say ‘Comic Sans MS', the rendering will change as well.
But if I change the default font-family to a non-system Font (those I installed later), the rending will not change at all, regardless of the font format.

Hopefully my findings could be of any help.

Coding Wishes,
Sylvester

**Comment by PeterWone** [comment]
> Similar problem here, only I set my custom typeface in the VS Code but what's printed is the default Menlo. Suppose this problem is either related to Non-System typefaces or when the fonts have format of .otf

It is almost a certainty that this or something similar is the problem. The browser must be able to resolve the typeface from the name. This hasn't been an issue for symbolic names like `serif` or `monospace` and on Windows where most people use it, the typeface name reported by VSCode settings has been correctly resolved by all common browsers, but this issue suggests this does not hold for MacOS. 

I do have a Mac for the specific purpose of debugging this kind of problem so I'll set it up and experiment with different fonts. It may be necessary to invent some kind of setting for mapping names if there isn't a single mapping function.

If you're interested in investigating this then you could save the page and supporting resources to a folder (I assume this is possible with Safari, certainly Chrome derivatives can do it). Then you can tinker with the font-family in the CSS until you figure out what to do to the name to get it to resolve.

**Comment by Xylopyrographer** [comment]
Hi and thanks for the extension. Just confirming this is still an issue.

macOS: 15.2

VSCode:
    Version: 1.96.4
    Commit: cd4ee3b1c348a13bafd8f9ad8060705f6d4b9cba
    Date: 2025-01-16T00:16:19.038Z
    Electron: 32.2.6
    ElectronBuildId: 10629634
    Chromium: 128.0.6613.186
    Node.js: 20.18.1
    V8: 12.8.374.38-electron.0
    OS: Darwin x64 24.2.0

Browser: Safari Version 18.2 (20620.1.16.11.8)

In my configuration, IDE typeface is JetBrains Mono, ligature enabled (though bug still exists with this set to disabled).

Not sure I followed all the CSS debugging hints above but glad to provide any information to help resolve.

But seriously Microsoft. No built-in Print capability???

**Comment by PeterWone** [comment]
> Hi and thanks for the extension. Just confirming this is still an issue.
> 
> macOS: 15.2
> 
> VSCode: Version: 1.96.4 Commit: cd4ee3b1c348a13bafd8f9ad8060705f6d4b9cba Date: 2025-01-16T00:16:19.038Z Electron: 32.2.6 ElectronBuildId: 10629634 Chromium: 128.0.6613.186 Node.js: 20.18.1 V8: 12.8.374.38-electron.0 OS: Darwin x64 24.2.0
> 
> Browser: Safari Version 18.2 (20620.1.16.11.8)
> 
> In my configuration, IDE typeface is JetBrains Mono, ligature enabled (though bug still exists with this set to disabled).
> 
> Not sure I followed all the CSS debugging hints above but glad to provide any information to help resolve.
> 
> But seriously Microsoft. No built-in Print capability???

Just to confirm, you specified `'JetBrains Mono'` in the setting `Editor: Font Family` and it renders properly in your editor?

Would you mind supplying screen snaps of the setting, the font working in the editor and not working in the preview? I can't produce this on my test Mac and I need to supply evidence for the Stack Overflow question I've posted. I loathe using Stack Overflow, the question itself has already been downvoted. No comment, no answer, just a kick in the teeth.

One thing to check, try deliberately using the wrong case in the font name to see whether it continues to work properly in the editor. If the editor is sensitive to font-name case this allows us to validate the correctness of the specified name.

**Comment by Xylopyrographer** [comment]
@PeterWone Screenshots below:
Did a bit of experimenting. Seems if I pick a font located in `/Library/Fonts/` it does not render with the Print extension. However, if I pick a font located in `/System/Library/Fonts/`, it renders as expected.

If I mis-name a font in Editor:Settings, VSC uses its default font.

I tried moving the font into ``/System/Library/Fonts/` but the OS won't allow that.

LEt me know if I can provide anything else.

Should note this was done using `pdconsec.vscode-print v0.14.1`

![Image](https://github.com/user-attachments/assets/414215fd-b07a-44c0-82e3-1fe756366ba3)
![Image](https://github.com/user-attachments/assets/f455faf1-35ee-4463-90be-9a2c7b623a64)
![Image](https://github.com/user-attachments/assets/3d22a199-1efd-413b-8971-d5923c8b6a83)

**Comment by PeterWone** [comment]
I notice that in your snap `JetBrains Mono` contains a space but is not quoted. Try quoting it like so `'JetBrains Mono'` and tell me whether that makes any difference. Mine looks like this:

![Image](https://github.com/user-attachments/assets/7f4ad3db-4375-408d-bf65-c7f61be1dc6b)

**Comment by Xylopyrographer** [comment]
OK, ran a few more tests.

On my system, only if the font exists in the directory `/System/Library/Fonts/`, including subdirectories in that folder, is it used by the Print extension.

And Print will use the first font in the VSCode Settings list that is located in that folder (and its subdirectories), ignoring those first in the list that are located elsewhere.

For example, with the following set in VSCode:

<img width="434" alt="Image" src="https://github.com/user-attachments/assets/246346bf-d943-432b-95a3-16d9599c9890" />

Marker Felt which is located at `/System/Library/Fonts/MarkerFelt.ttc` is used to render the print.

If I swap the order to this:
<img width="431" alt="Image" src="https://github.com/user-attachments/assets/b555cba8-c4de-4f41-aa9a-1b7a4b7856ef" />

then American Typewriter, located at `/System/Library/Fonts/Supplemental/AmericanTypewriter.ttc` is used.

In all cases, `Jetbrains Mono`, located at `/Library/Fonts/JetBrainsMono-Regular.ttf` is correctly used as the Editor font.

I also tried with `.ttf` files located in `/System/Library/Fonts/`, to eliminate `.ttf` vs `.ttc` as a potential issue. Results were the same.

Also, FWIW, unclear if the CSS's used by Print render ligatures? Though they are turned on in the VSCode settings, they were not rendered, but that may be because none of the fonts in `/System/Library/Fonts/` support ligatures?

Curious to see if you changed the order in your VSCode settings to match what I did if you see the same results as I'm wondering if `mono` isn't actually being used when rendering? I also tried the above replacing `'Jetbrains Mono'` with `'Fira Code'` and again the results were the same.

As a workaround, I've set things as:

<img width="427" alt="Image" src="https://github.com/user-attachments/assets/cb5e2717-44d9-4dcf-8fd8-54e8b7697a23" />


Annnnd one more thing; quoting font names with spaces does not seem to be required as:

<img width="428" alt="Image" src="https://github.com/user-attachments/assets/42fa8161-afe6-4235-b8b9-de5c4d981bc8" />

also renders using American Typewriter.

Let me know if I can do more digging.

**Comment by PeterWone** [comment]
Today I tried to reproduce the issue on a Mac. I installed Fira Code using the instructions provided by Bing Copilot, and... everything just worked. Here's a copy of my query and the resultant instructions. I downloaded the font from GitHub and installed it using Font Book. I did not specify which of the various version (TTF, WOFF etc) was to be used, I let Font Book decide all that. Then I started VS Code, found some source code and tried a preview, which used the font and rendered ligatures. 

### Please work the exercise. If Fira Code works, try installing JetBrains Mono using Font Book and see whether it sets up some kind of registration info that supports the magic of name resolution.

_A user reports that fonts specified in CSS font-family do not work on Mac unless they are system fonts. I cannot reproduce this because the Mac I use for testing only has system fonts. Tell me where I can download a free dev font like Fira Code and how to install it on a Mac._

## You can download Fira Code from the following sources:
 
 1. Google Fonts: Fira Code on Google Fonts
 2. GitHub: [Fira Code on GitHub](https://github.com/tonsky/FiraCode)

## To install Fira Code on your Mac, follow these steps:

1. **Download the Font:** Download the font file (usually a .ttf or .otf file) from one of the sources above.
2. **Open Font Book:** Open the Font Book app on your Mac. You can find it in your Applications folder or use Spotlight to search for it.
3. **Install the Font:** In Font Book, go to **File > Add Fonts**. Navigate to the location where you downloaded the font file, select it, and click **Open**.
4. **Validate the Font:** Font Book will automatically validate the font for any errors. If there are no issues, the font will be installed and ready to use.

Once installed, you should be able to use Fira Code in your CSS font-family declarations.

Does this help with your testing setup?

**Comment by PeterWone** [comment]
Another thing, I just noticed you write of `ttc` files rather than `ttf` so I looked up `ttc` in the context of Mac fonts. Apparently a TTC file is a "TrueType collection". I had a look at what's in the Fira Code download bundle. There's no TTC files so maybe this is related. 

I then searched for JetBrains Mono, found it on GitHub, downloaded the font package and same story. Works perfectly for me. Interestingly, there was no `ttc` in the package. The other listed installation method involved a tool called "brew" that I didn't have. I wonder whether the brew installation method is problematic.

**Comment by Xylopyrographer** [comment]
Peter: Long story short. Seems this is an issue with the macOS Safari browser. With the Print extension in the default state, it uses Safari. If I examine the Print preview page using the Safari Developer tools I get as per the image below.

![Image](https://github.com/user-attachments/assets/cea712f8-d1b9-4c33-b41b-2d5026b0a46c)

I can edit the fonts used, but regardless Safari will only let you select fonts located in the `/System/Library/Fonts/` folder. Those are the only choices in the dropdown menu. All fonts prior that are not in this folder are ignored.

Now if I  copy the preview page link from Safari and open Firefox and paste in the URL, it renders using JetBrains Mono. So sucess there.

However... if I open the Print extension preview and enter the path to Firefox as the alternate. browser, it does not stick. Safari is always used.

![Image](https://github.com/user-attachments/assets/d2376856-6b45-44f1-9cec-a05179f0430d)

Hope this helps progress the issue.

Note, .ttc .woff,  etc files: no change to any of the above.

**Comment by PeterWone** [comment]
Did you also tick `Use the alternate browser`? It's done like that so you can turn it on and off without losing the path

![Image](https://github.com/user-attachments/assets/254b1345-99d0-41eb-a1fd-1afcbf1d2a94)

**Comment by Xylopyrographer** [comment]
Ah! Misssed that bit.

Ticking that box and hitting the Print Preview icon now throws:
`Command failed: /Applications/Firefox.app http://localhost:62805/a857a55c-5fab-420b-865d-58a5a8b3278a/ /bin/sh: /Applications/Firefox.app: is a directory`

Which technically is true of all macOS applications, it's just the way Apple bundles all the resources of the app.

Changing the path to `/Applications/Firefox` also fails with:
`Command failed: /Applications/Firefox http://localhost:62805/6efbac03-c5a0-4985-a7f9-2c3ecd281f66/ /bin/sh: /Applications/Firefox: No such file or directory`.

**Comment by PeterWone** [comment]
You must supply the path to the executable. It's building a bash command to launch Firefox with a url argument. Look in that Firefox directory for the name of the executable. You can test the path in terminal, when you get it right you can use it to start Firefox from the command line.

On my Mac it appears that the path for Microsoft Edge is 

`"/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge"`

In Terminal, the command `"/Applications/Microsoft Edge.app/Contents/MacOS/Microsoft Edge" https://www.google.com` will launch Edge and open the google search engine page.

For Firefox your executable path will probably be something like `"/Applications/Firefox.app/Contents/MacOS/Firefox"`

**Comment by Xylopyrographer** [comment]
Changing the path to `/Applications/Firefox.app/Contents/MacOS/firefox` does work in that **_Firefox_** (version 134.0.2 (64-bit)) is launched and the Print preview page is displayed in the font selected in the VSCode editor. However VSCode reports the following error:
`[ERROR glean_core::metrics::ping] Invalid reason code active for ping usage-reporting 2025-02-03 11:49:31.879 firefox[42736:1918891] +[IMKClient subclass]: chose IMKClient_Modern 2025-02-03 11:49:31.879 firefox[42736:1918891] +[IMKInputSession subclass]: chose IMKInputSession_Modern`

Noting as well that with this, the Firefox window opens behind VSCode. Diffing behaviour than with Safari, which becomes the frontmost window when the Preview icon is clicked.

Trying with _**MS Edge**_ (Version 132.0.2957.127 (Official build) (x86_64)) works, and it becomes the foreground app but VSCode throws:
`Trying to load the allocator multiple times. This is *not* supported. 2025-02-03 12:01:29.859 Microsoft Edge[43301:1939276] +[IMKClient subclass]: chose IMKClient_Modern 2025-02-03 12:01:29.859 Microsoft Edge[43301:1939276] +[IMKInputSession subclass]: chose IMKInputSession_Modern`

Trying with **_Google Chrome_** (Version 132.0.6834.160 (Official Build) (x86_64)) works. Proper font used, Chrome comes to the foreground, no error messages in VSCode, even when changing the source file. Winner winner chicken dinner.

Obviosly all browsers are not alike.

So, the workarounds at the moment seem to be:
1. If wanting Print to use the same font as the VSCode Editor, experiment with different browsers other than Safari and then explicitly pick one in Print settings,
2. If wanting to use Safari, add a font in the VSCode Editor settings that resides in `/System/Library/Fonts/` (including it's sub folders) that sort of matches the font used in the Editor — noting that ligatures will not be there as (at this writing) none of the `.../System/...` fonts support ligatures.

Might be useful to add to the documentation for macOS a note that the complete path (inside the app package) must be used.

**Reply by PeterWone** [reply]
> Might be useful to add to the documentation for macOS a note that the complete path (inside the app package) must be used.

Nah I'll alter path resolution as follow
- On Mac if the supplied path resolves to a folder ending in `.app` then use `plist` (a npm package I discovered last night for doing this) to determine the executable path  
- On Windows if the supplied path resolves to a folder that contains only one exe then use that
- On Linux hell if I know I'll have to take a look at the way they package stuff

**Comment by PeterWone** [comment]
Because I use Edge without incident on my Mac I checked the settings. Apparently I have set Edge as the default browser, and when launched that way it works fine.

I will have to do further testing.

You said there was an exception reported launching Edge as the alternate browser. That doesn't happen on mine. I appear to have Monterey 12.7.6 and Edge 132.0.2975.127

Will install FF and try that but right now it's time to connect to the office and start work.

**Comment by PeterWone** [comment]
I researched this error
```
Trying to load the allocator multiple times. This is *not* supported. 2025-02-03 12:01:29.859 Microsoft Edge[43301:1939276] +[IMKClient subclass]: chose IMKClient_Modern 2025-02-03 12:01:29.859 Microsoft Edge[43301:1939276] +[IMKInputSession subclass]: chose IMKInputSession_Modern
```

This is an Edge error, apparently. It's not occurring on my system at all so maybe your Edge installation is somehow borked? Or maybe you have a happier combination of OS and browser.

I looked at the Glean exception for FF and one of the proposed fixes is to clear your cookies and cache. FFS, really Firefox? I actually hope that _doesn't_ work.

**Comment by PeterWone** [comment]
I Installed Firefox and it worked without incident. 

Then I tried Safari. Safari doesn't work _at all_ as an alternate browser. It launches but it assumes that when launched directly the argument must resolve in the filesystem and it prefixes the argument with `file:////` which obviously isn't going to work. Launched as the default browser it works but only with system fonts, exactly as you reported.

I will update the documentation to politely explain that Safari is a borked piece of shit. I will also mention that it is possible to experience the Firefox or Microsoft Edge glitches you described. Would you mind reinstalling them to see whether that resolves things please? If it does, I will include that as remedial advice.

Finally, Firefox is a more compliant browser anyway so I think it's probably appropriate to list the shortcomings of Safari and then suggest installing Firefox and setting it as the default browser (or Chrome, for people who don't mind Google spyware).

---

I think I know what's going on with Safari and fonts. I do not provide `@font-face` in the CSS because I'm not supplying the fonts. They're obviously already installed, or you couldn't be using them in the editor, and given I'm not supplying them, I don't _know_ where they are. I think Safari makes the unreasonable assumption that anything that isn't a default system font will be resolved to a file by `@font-face`.

Many of the devs on the VS Code team use Mac and vscode.dev is their love-child getting far more resource and dev time than it really deserves, so out of curiosity I exercised locally installed Fira Code in Safari. Our little Safari font problem also occurs for the web version of VS Code. It's probably not fixable. Well... Apple could fix it but that would require an admission that an Apple product was inferior to absolutely everything else on the market.

**Reply by Xylopyrographer** [reply]
> ...would require an admission that an Apple product was inferior to absolutely everything else on the market 🤣

• OK. Extinguished all traces of Firefox from my system and reinstalled, and after doing all the startup things it requires, quit the app. Firefox version is 135.0 (64-bit).

• Launched VSC, reset Firefox as the alternate browser for Print to use.

• Clicked the Print preview icon in the Editor. Firefox launched but did not become the foreground app. Brought it to the foreground and the preview was as expected 👍. The preview page also updates as changes are made to the source file in VSC.

• Closed all windows in Firefox, but left the app running. Clicked on the Print preview icon. Firefox pops to the foreground with the preview shown as it should. All good and the preview page also updates as changes are made to the source file in VSC.👍👍.

• Did see VSC throw one instance of:

`2025-02-04 15:17:41.732 firefox[63047:3272936] +[IMKClient subclass]: chose IMKClient_Modern 2025-02-04 15:17:41.732 firefox[63047:3272936] +[IMKInputSession subclass]: chose IMKInputSession_Modern`

but I think it has to do with timing. If I close the preview tab in Firefox while Print is still "talking" to it Print throws the error. Only saw it once so 🤷🏼‍♂️.

Seems then that things are sorted as much as they can be at the moment.

Thanks for the extension and for working through the issues. Much appreciated!

**Reply by PeterWone** [reply]
Incidentally I got the code working for automatically finding the executable from an OSX app manifest, it will appear in a later release after I get it to work for all platforms. Windows and Linux don't have a manifest like that so it's harder to support a folder path. Or maybe I won't, because users of those platforms don't expect it. Still thinking on that.

**Comment by PeterWone** [comment, marked as answer]
# Summary answer

- The font problem is a Safari problem resolved by using any other browser. We recommend Firefox. Edge and Chrome also work well but that's up to you. Evil empires, y'know. 
- If you get weird errors when Print launches your browser, try removing and re-installing the browser. This isn't a Print problem, it's a browser launch problem.
